backend/controller/cronjobs/cronjobs.go

```diff
@@ -0,0 +1,483 @@
+
+type Service struct {
+	config    Config
+	key       model.ControllerKey
+	originURL *url.URL
+
+	dal       DAL
+	scheduler Scheduler
+	call      ExecuteCallFunc
+
+	clock      clock.Clock
+	jobChanges *pubsub.Topic[jobChange]
+
+	hashRingState atomic.Value[*hashRingState]
+}
+
+func New(ctx context.Context, key model.ControllerKey, originURL *url.URL, config Config, dal DAL, scheduler Scheduler, call ExecuteCallFunc) *Service {
+	return NewForTesting(ctx, key, originURL, config, dal, scheduler, call, clock.New())
+}
+
+func NewForTesting(ctx context.Context, key model.ControllerKey, originURL *url.URL, config Config, dal DAL, scheduler Scheduler, call ExecuteCallFunc, clock clock.Clock) *Service {
+	svc := &Service{
+		config:     config,
+		key:        key,
+		originURL:  originURL,
+		dal:        dal,
+		scheduler:  scheduler,
+		call:       call,
+		clock:      clock,
+		jobChanges: pubsub.New[jobChange](),
+	}
+	svc.UpdatedControllerList(ctx, nil)
+
+	svc.scheduler.Parallel(backoff.Backoff{Min: time.Second, Max: jobResetInterval}, svc.resetJobs)
+	svc.scheduler.Singleton(backoff.Backoff{Min: time.Second, Max: time.Minute}, svc.killOldJobs)
+
+	go svc.watchForUpdates(ctx)
+
+	return svc
+}
+
+func (s *Service) NewCronJobsForModule(ctx context.Context, module *schemapb.Module) (jobs []dal.CronJob, err error) {
+	logger := log.FromContext(ctx)
+
+	start := s.clock.Now().UTC()
+	newJobs := []dal.CronJob{}
+	for _, decl := range module.Decls {
+		if verb, ok := decl.Value.(*schemapb.Decl_Verb); ok {
+			for _, metadata := range verb.Verb.Metadata {
+				if cronMetadata, ok := metadata.Value.(*schemapb.Metadata_CronJob); ok {
+					cronStr := cronMetadata.CronJob.Cron
+					schedule, err := cron.Parse(cronStr)
+					if err != nil {
+						logger.Errorf(err, "failed to parse cron schedule %q", cronStr)
+						continue
+					}
+					next, err := cron.NextAfter(schedule, start, false)
+					if err != nil {
+						logger.Errorf(err, "failed to calculate next execution for cron job %v:%v with schedule %q", module.Name, verb.Verb.Name, schedule)
+						continue
+					}
+					newJobs = append(newJobs, dal.CronJob{
+						Key:           model.NewCronJobKey(module.Name, verb.Verb.Name),
+						Ref:           schema.Ref{Module: module.Name, Name: verb.Verb.Name},
+						Schedule:      cronStr,
+						StartTime:     start,
+						NextExecution: next,
+						State:         dal.JobStateIdle,
+						// DeploymentKey: Filled in by DAL
+					})
+				}
+			}
+		}
+	}
+	return newJobs, nil
+}
+
+func (s *Service) CreatedOrReplacedDeloyment(ctx context.Context, newDeploymentKey model.DeploymentKey) {
+	// Rather than finding old/new cron jobs and updating our state, we can just reset the list of jobs
+	_ = s.resetJobsWithNewDeploymentKey(ctx, optional.Some(newDeploymentKey))
+}
+
+// resetJobs is run periodically via a scheduled task
```

Comment:
I found the word "reset" a bit confusing here and below. It looks more like it should be called "sync", as it's syncing the current set of jobs from the database?